ecosystem/ffmpeg_plugin/kahawai_enc.c

```diff
@@ -0,0 +1,396 @@
+        }
+    }
+
+    return 0;
+}
+
+static int kahawai_write_header(AVFormatContext *ctx) {
+  KahawaiMuxerContext *s = ctx->priv_data;
+  AVStream *st = NULL;
+
+  int packet_size = 0;
+
+  int ret = 0;
+
+  struct mtl_init_params param;
+  struct st20p_tx_ops ops_tx;
+
+  av_log(ctx, AV_LOG_VERBOSE, "kahawai_write_header triggered\n");
+
+  memset(&param, 0, sizeof(param));
+  memset(&ops_tx, 0, sizeof(ops_tx));
+
+  if ((NULL == s->port) || (strlen(s->port) > MTL_PORT_MAX_LEN)) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid port info\n");
+    return AVERROR(EINVAL);
+  }
+  param.num_ports = 1;
+  strncpy(param.port[MTL_PORT_P], s->port, MTL_PORT_MAX_LEN);
+  ops_tx.port.num_port = 1;
+  strncpy(ops_tx.port.port[MTL_PORT_P], s->port, MTL_PORT_MAX_LEN);
+
+  if (NULL == s->local_addr) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid local IP address\n");
+    return AVERROR(EINVAL);
+  } else if (sscanf(s->local_addr, "%hhu.%hhu.%hhu.%hhu",
+                    &param.sip_addr[MTL_PORT_P][0],
+                    &param.sip_addr[MTL_PORT_P][1],
+                    &param.sip_addr[MTL_PORT_P][2],
+                    &param.sip_addr[MTL_PORT_P][3]) != MTL_IP_ADDR_LEN) {
+    av_log(ctx, AV_LOG_ERROR, "Failed to parse local IP address: %s\n",
+           s->local_addr);
+    return AVERROR(EINVAL);
+  }
+
+  param.tx_sessions_cnt_max = s->session_cnt;
+  param.rx_sessions_cnt_max = 0;
+  param.flags = MTL_FLAG_BIND_NUMA | MTL_FLAG_DEV_AUTO_START_STOP; // FIXME
+  param.log_level = MTL_LOG_LEVEL_DEBUG; // log level. ERROR, INFO, WARNING
+  param.priv = NULL;                     // usr crx pointer
+  param.ptp_get_time_fn = NULL;
+  param.lcores = NULL;
+
+  if (NULL == s->dst_addr) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid destination IP address\n");
+    return AVERROR(EINVAL);
+  } else if (sscanf(s->dst_addr, "%hhu.%hhu.%hhu.%hhu",
+                    &ops_tx.port.dip_addr[MTL_PORT_P][0],
+                    &ops_tx.port.dip_addr[MTL_PORT_P][1],
+                    &ops_tx.port.dip_addr[MTL_PORT_P][2],
+                    &ops_tx.port.dip_addr[MTL_PORT_P][3]) != MTL_IP_ADDR_LEN) {
+    av_log(ctx, AV_LOG_ERROR, "Failed to parse destination IP address: %s\n",
+           s->dst_addr);
+    return AVERROR(EINVAL);
+  }
+
+  if ((s->udp_port < 0) || (s->udp_port > 0xFFFF)) {
+    av_log(ctx, AV_LOG_ERROR, "Invalid UDP port: %d\n", s->udp_port);
+    return AVERROR(EINVAL);
+  }
+  ops_tx.port.udp_port[MTL_PORT_P] = s->udp_port;
+
+  ops_tx.width = s->width = ctx->streams[0]->codecpar->width;
+  ops_tx.height = s->height = ctx->streams[0]->codecpar->height;
+
+  s->pixel_format = ctx->streams[0]->codecpar->format;
+  if (s->pixel_format == AV_PIX_FMT_NONE) {
+    av_log(ctx, AV_LOG_ERROR, "No such pixel format: %s.\n",
+           av_pix_fmt_desc_get(s->pixel_format));
+    return AVERROR(EINVAL);
+  } else if (s->pixel_format != AV_PIX_FMT_RGB24) {
+    av_log(ctx, AV_LOG_ERROR, "Only rgb24 are supported\n");
+    return AVERROR(EINVAL);
```

Comment:
Would you please consider to add the support for yuv422-10bit-le as well, so that kahawai indev and outdev could be used together directly without csc?